Flexible DDL Expression Support

eakmanrq opened the pull request:
Prior to this change the DDL for tables were expected to follow a specific order in order to get transpiled correctly. In addition there was a distinction between "properties" and "options" that seemed to apply to MySQL only. 

This change updates the parser to treat all properties the same (no longer finds properties and then options). In addition it makes property specific expressions instead of a generic Property expression. It also preserved the order in which the properties were defined. Dialects can decide which properties should be at the root level, under the `WITH` clause, or under a `TBLPROPERTIES` clause. They can then override the sql generation behavior for each property if needed. The base implementation is configured to generate MySQL correctly since that is how the tests are defined. 

Better DDL support is still needed for Snowflake and BigQuery and I've decided to make that as out of scope for this PR.

tobymao commented on sqlglot/expressions.py:
```diff
@@ -1358,19 +1393,26 @@ def ctas(self, table, properties=None, dialect=None, parser_opts=None, copy=True
+        if properties:
+            properties_str = " ".join(
+                [
+                    f"{k} = {v}" if not isinstance(v, str) else f"{k} = '{v}'"
```
nit: reverse this so that the if statement is simpler if isinstance(str)...

tobymao commented on sqlglot/generator.py:
```diff
@@ -38,8 +38,18 @@ class Generator:
 
     TRANSFORMS = {
+        exp.AnonymousProperty: lambda self, e: self.property_sql(e),
+        exp.AutoIncrementProperty: lambda self, e: f"AUTO_INCREMENT={e.text('value')}",
```
should we do self.sql(e, 'value') instead of all the text references, and then the quotes vs no quote is automatic

eakmanrq replied:
Good call that helps a lot

tobymao commented on sqlglot/generator.py:
```diff
@@ -38,8 +38,18 @@ class Generator:
         exp.DateAdd: lambda self, e: f"DATE_ADD({self.sql(e, 'this')}, {self.sql(e, 'expression')}, {self.sql(e, 'unit')})",
         exp.DateDiff: lambda self, e: f"DATE_DIFF({self.sql(e, 'this')}, {self.sql(e, 'expression')})",
+        exp.EngineProperty: lambda self, e: f"ENGINE={e.text('value')}",
+        exp.FileFormatProperty: lambda self, e: f"format = '{e.text('value')}'",
```
looks like your spacing is not consistent? also should it be caps FORMAT

eakmanrq replied:
This has to do with the identity tests. They expect root expressions to have the space and lowercase while WITH expressions to have no space and uppercase. Do we want to change the identity tests to reflect a more consistent approach?

tobymao replied:
you can fix the tests yes

tobymao commented on sqlglot/generator.py:
```diff
@@ -394,7 +397,31 @@ def partition_sql(self, expression):
-        return self.properties("WITH", expression)
+        root_expression = expression.copy()
+        with_expression = expression.copy()
+        table_properties_expression = expression.copy()
```
i'm not a fan of how this was done.

we should only have root properties and with properties, tblproperties should only be overridden in hive, hive should overwrite with properties or something to be tblproperties

tobymao replied:
also, let's not copy expression n times and remove, can we just create a new root with the appropriate children?

eakmanrq replied:
So I do something like this now:
```
root_expression = exp.Properties(expressions=root_properties)
with_expression = exp.Properties(expressions=with_properties)
```

The risk I see though is that since we aren't cloning here we are losing any addition information that could be attached to properties that somewhere in later calls would assume exists. There is not an example of this today but wanted to point out that this could create some non-intuitive behavior. Another way to do this is to clone and then replace `expressions` with a list of properties. From your comment I think what I did is what you are wanting but I want to point out that other options too.

tobymao replied:
yea this is what i was looking for, not exactly sure in understanding the case you’re talking about though

tobymao commented on sqlglot/generator.py:
```diff
@@ -947,17 +974,17 @@ def format_time(self, expression):
         )
 
-    def expressions(self, expression, key=None, flat=False, indent=True):
+    def expressions(self, expression, key=None, flat=False, indent=True, item_sep=", "):
```
maybe just call it sep=", "

tobymao commented on sqlglot/parser.py:
```diff
@@ -572,93 +539,145 @@ def parse_option(option, token, option_lambda):
-
-    def _parse_properties(self, schema):
-        properties = []
+        if self._match(TokenType.USING):
```
can we move this all to a dictionary with a lambda

TokenType.USING: lambda ... 

that way every dialect can override how it reads property, also then it will be constant time to match

eakmanrq replied:
Yeah nice improvement.

tobymao commented on sqlglot/parser.py:
```diff
@@ -572,93 +539,145 @@ def parse_option(option, token, option_lambda):
+        if self._match(TokenType.USING):
+            return self.expression(
+                exp.TableFormatProperty,
+                this=exp.Literal.string(c.TABLE_FORMAT),
```
these constants were useful because we were using them on read too, but now that we have first class expressions, they are unnecessary, can you delete all the constants and just hardcode the default string here?

tobymao commented on sqlglot/parser.py:
```diff
@@ -572,93 +539,145 @@ def parse_option(option, token, option_lambda):
+            return self.expression(
+                exp.FileFormatProperty,
+                this=exp.Literal.string(c.FILE_FORMAT),
+                value=exp.Literal.string(self._parse_var().text("this")),
```
instead of text("this"), you can just call .name

tobymao commented on sqlglot/parser.py:
```diff
@@ -572,93 +539,145 @@ def parse_option(option, token, option_lambda):
+                this=exp.Literal.string(c.AUTO_INCREMENT),
+                value=self._parse_var() or self._parse_number(),
+            )
+        if self._match(TokenType.CHARACTER_SET) or self._match_sequence(
```
this is the only one that won't map in the dictionary, but that's ok, just have that one after the main dictionary matcher

tobymao commented on sqlglot/parser.py:
```diff
@@ -572,93 +539,145 @@ def parse_option(option, token, option_lambda):
+                this=exp.Literal.string(c.COLLATE),
+                value=self._parse_var() or self._parse_string(),
+            )
+        if self._match(TokenType.SCHEMA_COMMENT):
```
should all of these be match_sequence too? or are there dialects without mandatory eq?

eakmanrq replied:
I purposely did it this way to make the parsing more flexible. For example MySQL has `COMMENT=<comment>` while Spark has `COMMENT <comment>`. So I think this approach is more robust since we don't really care from a parsing perspective if the equal is there or not.

tobymao replied:
sounds good

tobymao commented on sqlglot/parser.py:
```diff
@@ -572,93 +539,145 @@ def parse_option(option, token, option_lambda):
+            )
+        if self._match_sequence(
+            TokenType.VAR, TokenType.EQ, advance=False
+        ) or self._match_sequence(TokenType.VAR, TokenType.EQ, advance=False):
```
what's this or statement both var eq? typo?

tobymao replied:
how come you're not advancing it? so that you can absorb the key? maybe match_sequence can return a tuple?

eakmanrq replied:
Yeah typo thanks for catching.

Yeah I want to absorb the key. Personally this pattern seems fine... with later Python releases we get the walrus operator which then combined with returning the tuple could be a nice solution.

tobymao replied:
ok

tobymao commented on sqlglot/parser.py:
```diff
@@ -1962,18 +1981,17 @@ def _match_set(self, types):
+            if not actual_token or sequence_token_type != actual_token.token_type:
+                return False
+        if advance:
+            self._advance(len(sequence))
```
you may actually be able to just do self._advance(i)

eakmanrq replied:
`self._advance(i+1)` technically works but IDE isn't happy with it. I think the problem is that in the case where sequence is empty then the i never gets initialized. That really shouldn't happen in practice but going to keep the current way since it seems comparable and removes a warning.

tobymao replied:
what if you set i = 0?

eakmanrq replied:
Yeah but now we are adding another line of code... seems all messier then this approach.

tobymao replied:
as long as there's no performance implication

tobymao replied:
i'm actually a bit afraid that _match_sequence is gonna significantly slow down the parser. you're now dynamically getting grabbing tokens instead of using the cache that was previously used 

self._curr and self._next

was this change necessary? do you even match more than just the pair?

eakmanrq replied:
This goes back to the typo that you mentioned. Originally I was matching on 3 things but then realized later I only needed 2. So yeah I can remove this now since it is not currently needed. Once someone wants to match on >2 items then we will need it.

tobymao replied:
cool yea let’s revert it for now. in general if you need to match things like that you can also use the tokenizer like with time zone

tobymao commented on sqlglot/parser.py:
```diff
@@ -443,12 +444,15 @@ def _find_token(self, token, sql):
 
         return index
 
+    def _get_token_at_index(self, index):
```
nit can you call this def _get_token(self, index)

tobymao commented on sqlglot/tokens.py:
```diff
@@ -383,6 +384,7 @@ class Tokenizer(metaclass=_Tokenizer):
         "FUNCTION": TokenType.FUNCTION,
         "FOLLOWING": TokenType.FOLLOWING,
         "FOREIGN KEY": TokenType.FOREIGN_KEY,
+        "FORMAT": TokenType.FORMAT,
```
whenever you add a token you need to add it to the ID_VAR_CONSTANTS or else you won't be able to use it as an alias 

select x format

eakmanrq replied:
Thanks!

tobymao commented on sqlglot/expressions.py:
```diff
@@ -759,6 +754,46 @@ class Property(Expression):
     arg_types = {"this": True, "value": True}
 
 
+class TableFormatProperty(Property):
```
do you like having all these with the property name on it? seems a bit unnecessary since the subclass is already property

so just TableFormat and PartitionedBy etc

eakmanrq replied:
Yeah I actually do like it. When someone is seeing these expressions referenced in code they don't see the class they inherited from. So it is nice making it explicit IMO. 

Also there is `AnonymousProperty` which really needs the property suffix to get it's purposes IMO. That also helps reinforce the point above.

tobymao replied:
sounsd good

tobymao commented on sqlglot/generator.py:
```diff
@@ -394,7 +402,27 @@ def partition_sql(self, expression):
+            p_class = p.__class__
+            if p_class in self.ROOT_PROPERTIES:
+                root_properties.append(p)
+            if p_class in self.WITH_PROPERTIES:
```
elif?

tobymao commented on sqlglot/dialects/presto.py:
```diff
@@ -151,6 +151,15 @@ class Parser(Parser):
     class Generator(Generator):
         STRUCT_DELIMITER = ("(", ")")
 
+        NATIVE_PROPERTIES = []
```
delete this?

eakmanrq replied:
Good catch thanks!

tobymao commented:
can you delete constants.py?

eakmanrq commented:
The file existed before this PR so I just removed the things that I added to it. It should be a small follow up PR if you want it completely removed but I see it as out of scope for this PR.
